### ISSUE TITLE
Terraform to manage the Cluster Security Group

### ISSUE THREAD

**Issue by adelwin** [terraform-aws-modules/terraform-aws-eks#1542]
## Description

Using the terraform module version 17.1.0 I can create the cluster, it came up just nice.
But it has this additional SG under `vpc_config`.
I've already tried to specify both `cluster_create_security_group` and `worker_create_security_group` to `false`.
But this SG is still created with default rule.
Is there a way to specify _not_ to create this SG?

The reason is because this SG is automatically created with a rule that is too wide for our enterprise use-case.
And it's currently being flagged.

This is the configuration that I used.
```hcl
module "eks" {
  source          = "terraform-aws-modules/eks/aws"
  version         = "17.1.0"
  tags            = merge(var.app_tags, local.common_tags)

  cluster_name    = local.eks_name
  cluster_version = var.k8s.k8s_version
  vpc_id          = data.aws_vpc.lz.id
  subnets         = data.aws_subnet.tier2.*.id

  cluster_enabled_log_types = var.cluster_log_types
  cluster_service_ipv4_cidr = var.k8s.cidr_service

  cluster_endpoint_private_access       = true
  cluster_endpoint_private_access_cidrs = distinct(concat([data.aws_vpc.lz.cidr_block], var.additional_private_cidr))

  cluster_endpoint_public_access        = false
  cluster_endpoint_public_access_cidrs  = distinct(concat(var.public_cidr, var.additional_public_cidr))

  manage_cluster_iam_resources          = true
  enable_irsa                           = true
  cluster_iam_role_name                 = local.eks_name
  workers_role_name                     = "node-${var.app_ref}"
  workers_additional_policies           = [
    "arn:aws:iam::aws:policy/CloudWatchLogsFullAccess",
    "arn:aws:iam::aws:policy/AmazonSSMManagedInstanceCore"
  ]

  node_groups_defaults                  = local.node_group_defaults
  node_groups                           = var.k8s.node_groups

  map_roles = local.common_mapped_roles
  map_users = local.common_mapped_users

  cluster_create_endpoint_private_access_sg_rule = false
  cluster_create_security_group = false
  cluster_security_group_id = aws_security_group.controlplane.id
  manage_worker_iam_resources = true

  worker_create_security_group = false
  worker_security_group_id = aws_security_group.workers.id
}
```

**Comment by bryantbiggs** [comment]
I don't know what is "too wide" but the `vpc_config` is required by EKS https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/eks_cluster#vpc_config-arguments 

https://github.com/terraform-aws-modules/terraform-aws-eks/blob/1d848b56bf7c43ce708bd7cdee45589f816e05f0/cluster.tf#L17-L23

**Comment by adelwin** [comment]
@daroga0002 which I did.
OK, let's take a step back first and align our terminology first.

When we create a cluster, there are 3 SG created.
* cluster security group
  * which we can override using `cluster_security_group_id`
  * which actually, eventually will be seen from the UI as "Additional Security Group"
```hcl
module "eks" {
  source          = "terraform-aws-modules/eks/aws"
  version         = "17.1.0"
  tags            = merge(var.app_tags, local.common_tags)

  cluster_name    = local.eks_name
  cluster_version = var.k8s.k8s_version

  cluster_create_security_group = false
  cluster_security_group_id = aws_security_group.controlplane.id
}
```

* worker security group (which we can override using `worker_security_group_id`)
  * which we can override using `worker_security_group_id`
  * which is not seen in the EKS configuration page, but this SG is attached to all the worker nodes
```hcl
module "eks" {
  source          = "terraform-aws-modules/eks/aws"
  version         = "17.1.0"
  tags            = merge(var.app_tags, local.common_tags)

  cluster_name    = local.eks_name
  cluster_version = var.k8s.k8s_version
 
  worker_create_security_group = false
  worker_security_group_id = aws_security_group.workers.id
}
```

* primary security group (this is the SG in question)
  * in the EKS configuration page it's actually shown as "Cluster security group"
  * we never specify this in the TF, but it's always created
  * we can add rules into it, but can't take out existing rules
  * one of the rules created out-of-the-box is an egress rule to 0.0.0.0/0 (which is my concern at the moment)

I want to take out this "primary security group".. or at the very least reduce the egress. 
Or perhaps I am not understanding the purpose of having this SG in the first place. 
If that's the case, please help to explain to me.

Thank you.

**Comment by daroga0002** [comment]
So I have taken a basic example from [here](https://github.com/terraform-aws-modules/terraform-aws-eks/tree/master/examples/basic) and replaced lines https://github.com/terraform-aws-modules/terraform-aws-eks/blob/a9b84558b1abb4ca2a4dd07f571439ccfff79f9d/examples/basic/main.tf#L103-L138 by following code:
```bash
module "eks" {
  source                        = "../.."
  cluster_name                  = local.cluster_name
  cluster_version               = "1.20"
  subnets                       = module.vpc.private_subnets
  vpc_id                        = module.vpc.vpc_id
  cluster_create_security_group = false
  cluster_security_group_id     = aws_security_group.controlplane.id
  worker_create_security_group  = false
  map_roles                     = var.map_roles
  map_users                     = var.map_users
  map_accounts                  = var.map_accounts
  manage_aws_auth               = false
}

resource "aws_security_group" "controlplane" {
  name        = "eks"
  description = "EKS controlplane own group"
  vpc_id      = module.vpc.vpc_id
}
```

Then I run terraform and in plan was only one security group (`resource aws_security_group.controlplane`), I applied it and this created me:
![image](https://user-images.githubusercontent.com/20227477/132464380-648e46ae-7c16-4419-bf12-e261ac3834d5.png)

So I have created new cluster from WebUI where I can specify security group only in one field so taken existing one:
![image](https://user-images.githubusercontent.com/20227477/132464576-d555444c-62d8-4495-bdaa-99de8db9d39a.png)

and thi created me cluster:
![image](https://user-images.githubusercontent.com/20227477/132464729-ee4a4ae4-86b3-466a-a4ab-27430d60ce8c.png)


this created me `eks-cluster-sg-***` security group automatically, so this brings us to assumption that this group is created by EKS AWS backend and we cannot influence it at all.

 Looking into docs:
 https://docs.aws.amazon.com/eks/latest/userguide/sec-group-reqs.html
 
 we have:
 ```
Amazon EKS clusters, starting with Kubernetes version 1.14 and platform version eks.3, create a cluster security group when they are created. This also happens when a cluster of an earlier version is upgraded to this Kubernetes version and platform version. A cluster security group is designed to allow all traffic from the control plane and managed node groups to flow freely between each other. By assigning the cluster security group to the elastic network interfaces created by Amazon EKS that allow the control plane to communicate with the managed node group instances, you don't need to configure complex security group rules to allow this communication. Any instance or network interface that is assigned this security group can freely communicate with other resources with this security group.
```

In same docs we have:
```
If you need to limit the open ports between the control plane and nodes, the default cluster security group can be modified to allow only the following required minimum ports.
```

but this must be modified after creation, so from terraform side we dont have such possibility to influence it.


Also I found https://github.com/aws/containers-roadmap/issues/1355 which is exactly what I have described above.



To summarise we dont have possibility to influence that security group from this module

**Comment by daroga0002** [comment]
adding reference to terraform provider issue https://github.com/hashicorp/terraform-provider-aws/issues/18856

**Comment by adelwin** [comment]
Thanks @daroga0002 for the clarification.

**Comment by lisfo4ka** [comment]
Maybe we can get this security group id with data source since we know its name pattern and attach a new egress security group rule which will overwrite the existing one. But I'm not sure if it wouldn't cause any conflict in the next terraform apply, this idea should be tested properly.

**Comment by adelwin** [comment]
@lisfo4ka,
I've tried that.. but since the SG works in additive manner, no matter what we set, it's not going to help since it's already open to the widest possible configuration.
And seems we can only "add" sg-rules, we can't actually delete sg-rules from terraform.
I noticed we can actually delete the sg-rules line from AWS Console UI tho.
But doing things that are highly dependant on sequence like that isn't very good

**Comment by shreeshsp** [comment]
@daroga0002  @adelwin 
Has terraform provided any ways of deleting sg-rules in their recent feature releases or still we don't have any way of deleting sg-rules using terraform ?

**Comment by daroga0002** [comment]
nothing changed as per my knowledge

**Comment by adelwin** [comment]
Not that I know of.
I mean I can understand the reasoning, because the sg-rules are not really created by TF per-se, it's created internally by AWS on creation of the cluster.
I believe AWS API should allow for full-override.

**Comment by ArtunSubasi** [comment]
I just stumbled across this issue. Our project currently has the same issue with the default outgoing security group rule in the automatically created cluster security group. This blocks our infrastructure automation as the AWS API does not provide a way to override the default rules in the cluster security group.

I opened a AWS support ticket using premium support and asked for an extension in the CreateCluster API. I'll let you know about about the reaction. Even if they agree, it will take a while before it is integrated into the AWS API and then to the Terraform EKS module. But hey, we have to start somewhere. :)

**Comment by roman-chudik-codeit** [comment]
> I just stumbled across this issue. Our project currently has the same issue with the default outgoing security group rule in the automatically created cluster security group. This blocks our infrastructure automation as the AWS API does not provide a way to override the default rules in the cluster security group.
> 
> I opened a AWS support ticket using premium support and asked for an extension in the CreateCluster API. I'll let you know about about the reaction. Even if they agree, it will take a while before it is integrated into the AWS API and then to the Terraform EKS module. But hey, we have to start somewhere. :)

Hi @ArtunSubasi, is there any updates in that from AWS support?

**Comment by oliverpark999** [comment]
I have the same problem.
If you write the code below and build Infra,
```
vpc_config {
security_group_ids = [aws_security_group.plane-sg.id]
subnet_ids =
}
```
The EKS Cluster is assigned an Auto-Generated SG.
Then I have no choice but to manage SG by [manual], not Terraform.

**Comment by bryantbiggs** [comment]
The cluster primary security group is created by the EKS service itself. This is not a module issue
- https://docs.aws.amazon.com/eks/latest/userguide/sec-group-reqs.html
- https://github.com/aws/containers-roadmap/issues/1355

**Comment by fartosh** [comment]
@bryantbiggs I understand that we cannot modify the security groups that are created by EKS by default. However, is there any way to force node groups to use the security groups created by the module?

**Comment by bryantbiggs** [comment]
Which security group exactly?

**Comment by fartosh** [comment]
The code related is:
```
module "eks" {
  source = "github.com/terraform-aws-modules/terraform-aws-eks?ref=v18.28.0"

  cluster_name                    = "example"
  cluster_version                 = "1.23"
  cluster_endpoint_private_access = true
  cluster_endpoint_public_access  = false

  eks_managed_node_groups = {
    coredns = {
      desired_size                 = 1
      instance_types               = ["t3.micro"]
      create_launch_template       = false
      launch_template_name         = ""
      iam_role_additional_policies = ["arn:aws:iam::aws:policy/AmazonSSMManagedInstanceCore"]
    }
  }
}
```
and by default, both the `create_node_security_group` and `create_cluster_security_group` are set to `true`. Thus, the 2 security groups are created:

1. Name tag: `example-cluster`, security group name: `example-cluster-<timestamp>`
2. Name tag: `example-node`, security group name: `example-node-<timestamp>`

The first one is listed under `Additional security groups` in the Networking tab of the EKS console. The second one is not used anywhere.

So the question is, can I (and how) force the EKS cluster and the node group to attach the security groups created by the module instead of the default EKS security groups that do not pass security scanning because of the excessive egress rules.

**Comment by bryantbiggs** [comment]
> create_launch_template       = false
> launch_template_name         = ""

You disabled the use of the custom launch template so no, you don't have the ability to control security groups attached to the nodes through Terraform - the EKS Managed Node Group service will use its own default launch template and attach the cluster primary security group. If you want to use the security groups created by the module, then you need to use a custom launch template (which is the default behavior of the module)

**Comment by fartosh** [comment]
@bryantbiggs just deleting those 2 lines helped with the node group's security groups, thanks!
